### PR TITLE
Added support for group private key rotation

### PR DESCRIPTION
**Pull request by ToniaSchlick**

### CODE REVIEW

File: src/frame/sdk/GroupOperations.ts [hunk 1]
```diff
@@ -15,6 +15,27 @@ export function groupCreate(signingKeys: SigningKeyPair, memberList: UserOrGroup
     return WorkerMediator.sendMessage<WMT.GroupCreateWorkerResponse>(payload).map(({message}) => message);
 }
 
+/**
+ * Send message to the worker to rotate the group private key and encrypt the new private key to the provided list of admins
+ */
+export function rotateAndEncryptNewGroupPrivateKeyToList(
+    encryptedGroupKey: TransformedEncryptedMessage,
+    adminList: UserOrGroupPublicKey[],
+    myPrivateKey: PrivateKey<Uint8Array>,
```
Comment (ernieturner) [inline src/frame/sdk/GroupOperations.ts]: Maybe just `userPrivateMasterKey`

File: src/frame/worker/GroupCrypto.ts [hunk 2]
```diff
@@ -27,6 +27,36 @@ export function createGroup(signingKeys: SigningKeyPair, memberList: UserOrGroup
         })
         .errorMap((error) => new SDKError(error, ErrorCodes.GROUP_KEY_GENERATION_FAILURE));
 }
+/**
+ *  Rotate current group private key. Then encrypts rotated private to the groups admins.
+ */
+export function rotatePrivateKey(
+    encryptedGroupKey: TransformedEncryptedMessage,
+    adminList: UserOrGroupPublicKey[],
+    adminPrivateKey: PrivateKey<Uint8Array>,
```
Comment (ernieturner) [inline src/frame/worker/GroupCrypto.ts]: Ditto here from above with the name of this key

File: src/frame/worker/crypto/recrypt/RecryptWasm.ts [hunk 3]
```diff
@@ -125,13 +125,39 @@ const rotateUsersPrivateKey = (userPrivateKey: Uint8Array): Future<Error, {newPr
...
-
 export const rotateUsersPrivateKeyWithRetry = (userPrivateKey: Uint8Array): Future<Error, {newPrivateKey: Uint8Array; augmentationFactor: Uint8Array}> => {
     return rotateUsersPrivateKey(userPrivateKey).handleWith(() => rotateUsersPrivateKey(userPrivateKey));
 };
 
+/**
+ * Generates a new key pair and takes the private key as the new group private key, The existing private key is then subtracted from the newly
+ * generated private key as the augmentation factor.
+ */
+const rotateGroupPrivateKey = (ExistingGroupPrivateKey: Uint8Array): Future<Error, {newPrivateKey: Uint8Array; augmentationFactor: Uint8Array}> => {
```
Comment (ernieturner) [inline src/frame/worker/crypto/recrypt/RecryptWasm.ts]:
```suggestion
const rotateGroupPrivateKey = (existingGroupPrivateKey: Uint8Array): Future<Error, {newPrivateKey: Uint8Array; augmentationFactor: Uint8Array}> => {
```

File: src/frame/worker/crypto/recrypt/RecryptWasm.ts [hunk 4]
```diff
@@ -125,13 +125,39 @@ const rotateUsersPrivateKey = (userPrivateKey: Uint8Array): Future<Error, {newPr
...
+    return generateKeyPair().flatMap(({privateKey}) => {
+        if (isBufferAllZero(privateKey)) {
+            return Future.reject(new Error("Key rotation failed."));
+        }
+        const newPrivateKey = RecryptApi.hash256(privateKey);
```
Comment (ernieturner) [inline src/frame/worker/crypto/recrypt/RecryptWasm.ts]: Why is this necessary here to rotate a groups private key but it isn't necessary when we do user key rotation?

File: integration/components/groups/GroupDetail.tsx [hunk 5]
```diff
@@ -133,6 +137,40 @@ export default class GroupDetail extends React.Component<GroupDetailProps, Group
...
+    /**
+     * Verify the group ID the user typed is correct before making the actual delete call.
+     */
+    roatateGroupPrivateKey = () => {
+        IronWeb.group
+            .rotatePrivateKey(this.props.group.groupID)
+            .then((groupKeyUpdate) => {
+                logAction(`Group private key rotation was successful`);
+                this.setState({needsRotation: groupKeyUpdate.needsRotation});
+                this.getRotateGroupPrivateKeyIcon();
```
Comment (ernieturner) [inline integration/components/groups/GroupDetail.tsx]: There's no need to call this method here. It's not doing anything. With React, the `render()` function automatically gets called after you call `this.setState` so that the UI reflects the new state.

File: integration/components/groups/GroupDetail.tsx [hunk 6]
```diff
@@ -133,6 +137,40 @@ export default class GroupDetail extends React.Component<GroupDetailProps, Group
...
+                    <FloatingActionButton onClick={this.roatateGroupPrivateKey} mini backgroundColor={red700}>
+                        <ImageRotateRight />
+                    </FloatingActionButton>
+                );
+            } else
+                return (
+                    <FloatingActionButton onClick={this.roatateGroupPrivateKey} mini backgroundColor={grey400}>
+                        <ImageRotateRight />
+                    </FloatingActionButton>
+                );
```
Comment (ernieturner) [inline integration/components/groups/GroupDetail.tsx]: The formatting here is weird because you're using brackets for the if, but not for the else. Instead I'd just remove the else and do

```
if(){
   return..
}
return ..
```

File: integration/components/groups/NewGroup.tsx [hunk 7]
```diff
@@ -24,7 +25,8 @@ export default class NewGroup extends React.Component<NewGroupProps, NewGroupSta
     constructor(props: NewGroupProps) {
         super(props);
         this.state = {
-            isChecked: true,
+            isAddAsMemberChecked: true,
+            isNeedsRotationChecked: true,
```
Comment (ernieturner) [inline integration/components/groups/NewGroup.tsx]: This should default to false because that's the default in the SDK.

File: src/frame/sdk/GroupApi.ts [hunk 8]
```diff
@@ -145,6 +152,33 @@ export function create(
...
+            );
+        }
+        return UserApiEndpoints.callUserKeyListApi(group.adminIds)
+            .map((adminKeys) => adminKeys.result.map((user) => ({id: user.id, masterPublicKey: user.userMasterPublicKey})))
+            .flatMap((adminKeys) =>
+                GroupOperations.rotateGroupPrivateKeyAndEncryptToAdmins(group.encryptedPrivateKey, adminKeys, privateKey, ApiState.signingKeys())
+            )
+            .flatMap(({encryptedAccessKeys, augmentationFactor}) =>
+                GroupApiEndpoints.callGroupPrivateKeyUpdateApi(groupID, encryptedAccessKeys, augmentationFactor, group.currentKeyId).map((result) => ({
+                    needsRotation: result.needsRotation,
```
Comment (ernieturner) [inline src/frame/sdk/GroupApi.ts]: You can simplify this a little bit by destructing and doing 
```
.map(({needsRotation}) => ({needsRotation}))
```

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts [hunk 9]
```diff
@@ -52,6 +52,48 @@ describe("RecryptWasm", () => {
         });
     });
 
+    describe("rotateGroupPrivateKeyWithRetry", () => {
+        const groupPrivateKey = new Uint8Array([22, 33, 44]);
+        it("should result in an error when mocked generateKeyPair returns Uint8Array of zeros for new privateKey", () => {
+            jest.spyOn(Recrypt.getApi(), "generatePlaintext");
```
Comment (ernieturner) [inline src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: Should this spy be returning something? I looks like this test is meant to verify that if the plaintext we generate is all zeros it should fail, but it looks like here we're not generating anything? It should probably have a `.mockReturnValue(new Uint8Array(32))`.

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts [hunk 10]
```diff
@@ -52,6 +52,48 @@ describe("RecryptWasm", () => {
...
+            Recrypt.rotateGroupPrivateKeyWithRetry(groupPrivateKey).engage(
+                (error) => {
+                    expect(error.message).toEqual("Key rotation failed.");
```
Comment (ernieturner) [inline src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: I'd just remove this line. There's little to no benefit in having to update this test if we decide to change the text of the error message. The assert below this is sufficient.

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts [hunk 11]
```diff
@@ -52,6 +52,48 @@ describe("RecryptWasm", () => {
...
+                () => fail("Should not success when operation fails")
+            );
+        });
+
+        it("should result in an error when subtracting existing private key from the new private key", () => {
+            jest.spyOn(Recrypt.getApi(), "generatePlaintext").mockReturnValue(new Uint8Array([12, 23, 34]) as any);
+            jest.spyOn(MockRecrypt, "subtractPrivateKeys").mockReturnValue(new Uint8Array(32));
+            Recrypt.rotateGroupPrivateKeyWithRetry(groupPrivateKey).engage(
+                (error) => {
+                    expect(error.message).toEqual("Key rotation failed.");
```
Comment (ernieturner) [inline src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: Similar to above, it'd be better to have some assertion here about the calls to `generatePlaintext` or `subtractPrivateKeys` instead of asserting this exact error message text.

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts [hunk 12]
```diff
@@ -52,6 +52,48 @@ describe("RecryptWasm", () => {
...
+                },
+                () => fail("Should fail when private key subtraction results in zeroed private key")
+            );
+
+            expect(MockRecrypt.subtractPrivateKeys).toHaveBeenCalledTimes(2);
```
Comment (ernieturner) [inline src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: This needs to be moved up into the error callback.

File: src/frame/sdk/GroupApi.ts [hunk 13]
```diff
@@ -145,6 +152,33 @@ export function create(
...
+        if (!isGroupAdminResponse(group)) {
+            return Future.reject(
+                new SDKError(
+                    new Error("Current user is not authorized to rotate this group's private key as they are not a group administrator."),
+                    ErrorCodes.GROUP_PRIVATE_KEY_ROTATION_FAILURE
+                )
+            );
+        }
+        return UserApiEndpoints.callUserKeyListApi(group.adminIds)
+            .map((adminKeys) => adminKeys.result.map((user) => ({id: user.id, masterPublicKey: user.userMasterPublicKey})))
```
Comment (ernieturner) [inline src/frame/sdk/GroupApi.ts]: We should probably add a check here to make sure that we got back the same number of public keys that we requested. If there's 10 admins of this group and we only get 7 public keys back from this call, that should probably cause a failure of the rotation because the group membership won't be the same before/after key rotation.

Comment (coltfred) [reply src/frame/sdk/GroupApi.ts]: @ernieturner The server will reject if the admins list doesn't match what's in the database. You can still check if you want, but shouldn't ever happen.

Comment (ernieturner) [reply src/frame/sdk/GroupApi.ts]: Oh neat, good to know. Then ya, I'm fine with not having a check for that.

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts (hunk 9 above)
Comment (ToniaSchlick) [reply src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: In the rycrypt-wasm-binding  generatePlaintext is mocked for jest or other unit to return new Uint8Array(384). Do I still need to mock it here?

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts (hunk 10 above)
Comment (ToniaSchlick) [reply src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: done

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts (hunk 11 above)
Comment (ToniaSchlick) [reply src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: done

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts (hunk 12 above)
Comment (ToniaSchlick) [reply src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: done

File: src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts (hunk 9 above)
Comment (ernieturner) [reply src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: Oh right I forgot about the mock here. But when you re-mock it like you're doing here, it will override the other one. So I'm pretty sure it's not returning anything here. We should make it be explicit and have it return what you want based on what you're testing.

Comment (ToniaSchlick) [reply src/frame/worker/crypto/recrypt/tests/RecryptWasm.test.ts]: ok, done

File: integration/components/groups/GroupDetail.tsx [hunk 14]
```diff
@@ -133,6 +137,39 @@ export default class GroupDetail extends React.Component<GroupDetailProps, Group
...
+                return (
+                    <FloatingActionButton onClick={this.roatateGroupPrivateKey} mini backgroundColor={red700}>
+                        <ImageRotateRight />
+                    </FloatingActionButton>
+                );
+            }
+            return (
+                <FloatingActionButton onClick={this.roatateGroupPrivateKey} mini backgroundColor={grey400}>
+                    <ImageRotateRight />
+                </FloatingActionButton>
```
[the remark this answers is not in the document]
Comment (ToniaSchlick) [reply integration/components/groups/GroupDetail.tsx]: done
